src/Storages/MergeTree/MergeTreeIndexGin.cpp

```diff
@@ -249,18 +249,18 @@ bool MergeTreeIndexConditionGin::mayBeTrueOnGranuleInPart(MergeTreeIndexGranuleP
         throw Exception(ErrorCodes::LOGICAL_ERROR, "GinFilter index condition got a granule with the wrong type.");
 
     /// Check like in KeyCondition.
-    std::vector<BoolMask> rpn_stack;
+    std::vector<bool> rpn_stack;
     for (const auto & element : rpn)
     {
         if (element.function == RPNElement::FUNCTION_UNKNOWN)
         {
-            rpn_stack.emplace_back(true, true);
+            rpn_stack.emplace_back(true);
         }
         else if (element.function == RPNElement::FUNCTION_EQUALS
              || element.function == RPNElement::FUNCTION_NOT_EQUALS
              || element.function == RPNElement::FUNCTION_HAS)
         {
-            rpn_stack.emplace_back(granule->gin_filters[element.key_column].contains(*element.gin_filter, cache_store), true);
+            rpn_stack.emplace_back(granule->gin_filters[element.key_column].contains(*element.gin_filter, cache_store));
```

Comment:
In l. 266, l. 284 and l. 319,

```cpp
rpn_stack.back() = !rpn_stack.back(); /// (*)
```

calls this underneath:

```
BoolMask operator!() const { return {can_be_false, can_be_true}; }
```

Previously, (*) _always_ returned `true` because the code _always_ passed `true` as the 2nd argument of `BoolMask` (when emplacing an element).
(even though that is not always true, e.g. l. 341 sets `false` as 2nd argument)

Now, (*) negates whatever is already there.

Seems like we are changing semantics here, no?

Reply:
thanks for your comment! I reverted it back to the original implementation